# Part-1: Data Exploration and Preparation

Loading of necessary libraries

In [1]:
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.impute import KNNImputer
import seaborn as sns

Loading data

In [2]:
data = pd.read_csv('Integrated _CA1_Mushrooms Datasource.csv')
data

Mushroom ID Cap Shape Cap Color Gill Size Stalk Shape  Habitat  Edibility  \
0             1    sunken     brown    narrow    tapering   leaves     edible   
1             ?    convex     brown    narrow    tapering    urban     edible   
2             3      flat     brown     broad   enlarging  grasses     edible   
3             4    sunken      gray    narrow    tapering        ?  poisonous   
4             5      flat       red         ?   enlarging    urban     edible   
..          ...       ...       ...       ...         ...      ...        ...   
995         996    sunken       red     broad   enlarging    woods  poisonous   
996         997   conical     brown     broad    tapering    woods     edible   
997         998   conical         ?    narrow    tapering        ?     edible   
998         999    sunken       red    narrow   enlarging  grasses     edible   
999        1000    convex       red     broad    tapering   leaves          ?   

    Market Demand Spore Print Color Population  ... Stalk Surface Below Ring  \
0               1            orange   abundant  ...                    silky   
1               2             brown    several  ...                    scaly   
2               8            yellow  scattered  ...                        ?   
3               4            orange  scattered  ...                    scaly   
4               7             brown    several  ...                  fibrous   
..            ...               ...        ...  ...                      ...   
995             4             brown   abundant  ...                   smooth   
996             1            orange   numerous  ...                  fibrous   
997            10             brown  scattered  ...                    silky   
998             6             brown  scattered  ...                        ?   
999             9             black  scattered  ...                    scaly   

    Stalk Color Above Ring Stalk Color Below Ring  Veil Type Veil Color  \
0                    white                  brown  universal     orange   
1                     pink                  white  universal      white   
2                     gray                 orange    partial     orange   
3                     gray                 orange  universal     yellow   
4                   orange                  white    partial      white   
..                     ...                    ...        ...        ...   
995                  white                 orange  universal      white   
996                  white                  brown    partial      brown   
997                   gray                 orange    partial     yellow   
998                   pink                 orange  universal      brown   
999                  brown                      ?          ?          ?   

             Growth Rate Inspection Date Incubation date Temperature  \
0     0.9730964627139668        3/7/2021        3/2/2021          49   
1                      ?        3/5/2023        3/4/2023          13   
2     1.9776191125191191        1/8/2022        1/4/2022          10   
3     2.2727627620673605       2/10/2023        2/2/2023          22   
4     0.7188737847331463        1/9/2023        1/4/2023           9   
..                   ...             ...             ...         ...   
995    3.755729472021494        1/7/2022        1/5/2022          25   
996                    ?       5/11/2023        5/3/2023           4   
997  11.2583910697714988        3/7/2023        3/4/2023          19   
998          4.964073966        3/2/2023        3/2/2023          20   
999    4.451239984190497        4/3/2023        4/3/2023           0   

    Measurement  
0             C  
1             F  
2             C  
3             C  
4             C  
..          ...  
995           C  
996           C  
997           C  
998           C  
999           C  

[1000 rows x 24 columns]

Get statistical information of all columns

In [3]:
data.describe(include='all')


Mushroom ID Cap Shape Cap Color Gill Size Stalk Shape Habitat  \
count         1000      1000      1000      1000        1000    1000   
unique         879         6         6         3           3       6   
top              ?   conical      gray    narrow   enlarging   urban   
freq           122       188       193       464         491     198   
mean           NaN       NaN       NaN       NaN         NaN     NaN   
std            NaN       NaN       NaN       NaN         NaN     NaN   
min            NaN       NaN       NaN       NaN         NaN     NaN   
25%            NaN       NaN       NaN       NaN         NaN     NaN   
50%            NaN       NaN       NaN       NaN         NaN     NaN   
75%            NaN       NaN       NaN       NaN         NaN     NaN   
max            NaN       NaN       NaN       NaN         NaN     NaN   

       Edibility Market Demand Spore Print Color Population  ...  \
count       1000          1000              1000       1000  ...   
unique         3            11                 6          6  ...   
top       edible             3             brown   numerous  ...   
freq         446           109               202        199  ...   
mean         NaN           NaN               NaN        NaN  ...   
std          NaN           NaN               NaN        NaN  ...   
min          NaN           NaN               NaN        NaN  ...   
25%          NaN           NaN               NaN        NaN  ...   
50%          NaN           NaN               NaN        NaN  ...   
75%          NaN           NaN               NaN        NaN  ...   
max          NaN           NaN               NaN        NaN  ...   

       Stalk Surface Below Ring Stalk Color Above Ring Stalk Color Below Ring  \
count                      1000                   1000                   1000   
unique                        5                      6                      6   
top                       silky                   pink                   gray   
freq                        228                    190                    197   
mean                        NaN                    NaN                    NaN   
std                         NaN                    NaN                    NaN   
min                         NaN                    NaN                    NaN   
25%                         NaN                    NaN                    NaN   
50%                         NaN                    NaN                    NaN   
75%                         NaN                    NaN                    NaN   
max                         NaN                    NaN                    NaN   

        Veil Type Veil Color Growth Rate Inspection Date Incubation date  \
count        1000       1000        1000            1000            1000   
unique          3          5         920             168              87   
top     universal      white           ?       4/10/2023        5/1/2023   
freq          460        244          81              27              45   
mean          NaN        NaN         NaN             NaN             NaN   
std           NaN        NaN         NaN             NaN             NaN   
min           NaN        NaN         NaN             NaN             NaN   
25%           NaN        NaN         NaN             NaN             NaN   
50%           NaN        NaN         NaN             NaN             NaN   
75%           NaN        NaN         NaN             NaN             NaN   
max           NaN        NaN         NaN             NaN             NaN   

        Temperature Measurement  
count   1000.000000        1000  
unique          NaN           2  
top             NaN           C  
freq            NaN         967  
mean      14.666000         NaN  
std        8.698742         NaN  
min        0.000000         NaN  
25%        7.000000         NaN  
50%       15.000000         NaN  
75%       22.000000         NaN  
max       49.000000         NaN  

[11 rows x 24 columns]

Get data type information of all columns

In [4]:
data.dtypes

Mushroom ID                 object
Cap Shape                   object
Cap Color                   object
Gill Size                   object
Stalk Shape                 object
Habitat                     object
Edibility                   object
Market Demand               object
Spore Print Color           object
Population                  object
Odor                        object
Bruising                    object
Ring Type                   object
Stalk Surface Above Ring    object
Stalk Surface Below Ring    object
Stalk Color Above Ring      object
Stalk Color Below Ring      object
Veil Type                   object
Veil Color                  object
Growth Rate                 object
Inspection Date             object
Incubation date             object
Temperature                  int64
Measurement                 object
dtype: object

Remove the "Mushroom ID" column since it is irrelevant for to be input as machine learning model 

In [5]:
data = data.drop('Mushroom ID', axis=1)
data

Cap Shape Cap Color Gill Size Stalk Shape  Habitat  Edibility  \
0      sunken     brown    narrow    tapering   leaves     edible   
1      convex     brown    narrow    tapering    urban     edible   
2        flat     brown     broad   enlarging  grasses     edible   
3      sunken      gray    narrow    tapering        ?  poisonous   
4        flat       red         ?   enlarging    urban     edible   
..        ...       ...       ...         ...      ...        ...   
995    sunken       red     broad   enlarging    woods  poisonous   
996   conical     brown     broad    tapering    woods     edible   
997   conical         ?    narrow    tapering        ?     edible   
998    sunken       red    narrow   enlarging  grasses     edible   
999    convex       red     broad    tapering   leaves          ?   

    Market Demand Spore Print Color Population      Odor  ...  \
0               1            orange   abundant     musty  ...   
1               2             brown    several      foul  ...   
2               8            yellow  scattered     musty  ...   
3               4            orange  scattered  creosote  ...   
4               7             brown    several    almond  ...   
..            ...               ...        ...       ...  ...   
995             4             brown   abundant         ?  ...   
996             1            orange   numerous     musty  ...   
997            10             brown  scattered      foul  ...   
998             6             brown  scattered    almond  ...   
999             9             black  scattered      foul  ...   

    Stalk Surface Below Ring Stalk Color Above Ring Stalk Color Below Ring  \
0                      silky                  white                  brown   
1                      scaly                   pink                  white   
2                          ?                   gray                 orange   
3                      scaly                   gray                 orange   
4                    fibrous                 orange                  white   
..                       ...                    ...                    ...   
995                   smooth                  white                 orange   
996                  fibrous                  white                  brown   
997                    silky                   gray                 orange   
998                        ?                   pink                 orange   
999                    scaly                  brown                      ?   

     Veil Type Veil Color          Growth Rate Inspection Date  \
0    universal     orange   0.9730964627139668        3/7/2021   
1    universal      white                    ?        3/5/2023   
2      partial     orange   1.9776191125191191        1/8/2022   
3    universal     yellow   2.2727627620673605       2/10/2023   
4      partial      white   0.7188737847331463        1/9/2023   
..         ...        ...                  ...             ...   
995  universal      white    3.755729472021494        1/7/2022   
996    partial      brown                    ?       5/11/2023   
997    partial     yellow  11.2583910697714988        3/7/2023   
998  universal      brown          4.964073966        3/2/2023   
999          ?          ?    4.451239984190497        4/3/2023   

    Incubation date Temperature Measurement  
0          3/2/2021          49           C  
1          3/4/2023          13           F  
2          1/4/2022          10           C  
3          2/2/2023          22           C  
4          1/4/2023           9           C  
..              ...         ...         ...  
995        1/5/2022          25           C  
996        5/3/2023           4           C  
997        3/4/2023          19           C  
998        3/2/2023          20           C  
999        4/3/2023           0           C  

[1000 rows x 23 columns]

Replace '?' with NaN for proper handling

In [6]:
data = data.replace('?', pd.NA)
data

Cap Shape Cap Color Gill Size Stalk Shape  Habitat  Edibility  \
0      sunken     brown    narrow    tapering   leaves     edible   
1      convex     brown    narrow    tapering    urban     edible   
2        flat     brown     broad   enlarging  grasses     edible   
3      sunken      gray    narrow    tapering     <NA>  poisonous   
4        flat       red      <NA>   enlarging    urban     edible   
..        ...       ...       ...         ...      ...        ...   
995    sunken       red     broad   enlarging    woods  poisonous   
996   conical     brown     broad    tapering    woods     edible   
997   conical      <NA>    narrow    tapering     <NA>     edible   
998    sunken       red    narrow   enlarging  grasses     edible   
999    convex       red     broad    tapering   leaves       <NA>   

    Market Demand Spore Print Color Population      Odor  ...  \
0               1            orange   abundant     musty  ...   
1               2             brown    several      foul  ...   
2               8            yellow  scattered     musty  ...   
3               4            orange  scattered  creosote  ...   
4               7             brown    several    almond  ...   
..            ...               ...        ...       ...  ...   
995             4             brown   abundant      <NA>  ...   
996             1            orange   numerous     musty  ...   
997            10             brown  scattered      foul  ...   
998             6             brown  scattered    almond  ...   
999             9             black  scattered      foul  ...   

    Stalk Surface Below Ring Stalk Color Above Ring Stalk Color Below Ring  \
0                      silky                  white                  brown   
1                      scaly                   pink                  white   
2                       <NA>                   gray                 orange   
3                      scaly                   gray                 orange   
4                    fibrous                 orange                  white   
..                       ...                    ...                    ...   
995                   smooth                  white                 orange   
996                  fibrous                  white                  brown   
997                    silky                   gray                 orange   
998                     <NA>                   pink                 orange   
999                    scaly                  brown                   <NA>   

     Veil Type Veil Color          Growth Rate Inspection Date  \
0    universal     orange   0.9730964627139668        3/7/2021   
1    universal      white                 <NA>        3/5/2023   
2      partial     orange   1.9776191125191191        1/8/2022   
3    universal     yellow   2.2727627620673605       2/10/2023   
4      partial      white   0.7188737847331463        1/9/2023   
..         ...        ...                  ...             ...   
995  universal      white    3.755729472021494        1/7/2022   
996    partial      brown                 <NA>       5/11/2023   
997    partial     yellow  11.2583910697714988        3/7/2023   
998  universal      brown          4.964073966        3/2/2023   
999       <NA>       <NA>    4.451239984190497        4/3/2023   

    Incubation date Temperature Measurement  
0          3/2/2021          49           C  
1          3/4/2023          13           F  
2          1/4/2022          10           C  
3          2/2/2023          22           C  
4          1/4/2023           9           C  
..              ...         ...         ...  
995        1/5/2022          25           C  
996        5/3/2023           4           C  
997        3/4/2023          19           C  
998        3/2/2023          20           C  
999        4/3/2023           0           C  

[1000 rows x 23 columns]

Droping the rows that contains "?" in any of the rows

In [7]:
df_cleaned = data[~data[columns_with_question_mark].astype(str).apply(lambda x: x.str.contains('\?')).any(axis=1)].reset_index().drop('index', axis=1)
df_cleaned

Cap Shape Cap Color Gill Size Stalk Shape  Habitat  Edibility  \
0      sunken     brown    narrow    tapering   leaves     edible   
1        bell    yellow     broad   enlarging   leaves     edible   
2      convex       red    narrow   enlarging    urban     edible   
3      convex       red     broad    tapering    woods     edible   
4      convex     white    narrow   enlarging    urban     edible   
..        ...       ...       ...         ...      ...        ...   
126   conical     brown    narrow    tapering    urban  poisonous   
127      flat    yellow     broad   enlarging    urban     edible   
128   conical      gray    narrow   enlarging  grasses  poisonous   
129    sunken     white     broad   enlarging    woods  poisonous   
130    convex      gray     broad    tapering   leaves  poisonous   

    Market Demand Spore Print Color Population    Odor  ...  \
0               1            orange   abundant   musty  ...   
1               8             brown   numerous  almond  ...   
2               4            orange   numerous    foul  ...   
3               2            yellow   solitary    foul  ...   
4               1             white    several   musty  ...   
..            ...               ...        ...     ...  ...   
126             5            orange   solitary  almond  ...   
127             4             black   numerous   anise  ...   
128             4            yellow   solitary    foul  ...   
129             9            yellow   abundant    foul  ...   
130             3             black   numerous   anise  ...   

    Stalk Surface Below Ring Stalk Color Above Ring Stalk Color Below Ring  \
0                      silky                  white                  brown   
1                      silky                   gray                   pink   
2                      silky                 orange                 orange   
3                      scaly                  brown                   gray   
4                     smooth                   gray                  white   
..                       ...                    ...                    ...   
126                    scaly                  brown                   pink   
127                    silky                 orange                   pink   
128                   smooth                   gray                   gray   
129                    scaly                   gray                  brown   
130                   smooth                   gray                   gray   

     Veil Type Veil Color         Growth Rate Inspection Date Incubation date  \
0    universal     orange  0.9730964627139668        3/7/2021        3/2/2021   
1    universal     yellow         4.471316648        1/2/2022        1/2/2022   
2      partial      white  3.9783564690608615        3/7/2023        3/3/2023   
3      partial     yellow  1.5922072807712329       4/14/2023        4/5/2023   
4      partial     yellow   3.143234817538799        3/7/2023        3/3/2023   
..         ...        ...                 ...             ...             ...   
126    partial     orange  3.4035697795027953        5/7/2023        5/1/2023   
127    partial      brown  1.9385124867425854       3/11/2023        3/3/2023   
128    partial     yellow  1.4202926693977478        3/2/2023        3/2/2023   
129  universal     orange   3.779518165668043        1/7/2023        1/5/2023   
130    partial      white  1.5888757465434287        3/2/2022        3/1/2022   

    Temperature Measurement  
0            49           C  
1            18           C  
2             1           C  
3             5           C  
4             4           C  
..          ...         ...  
126          28           C  
127          28           C  
128          15           C  
129           6           C  
130           9           C  

[131 rows x 23 columns]

It can be observed that the cleaned dataset contains 131 rows where original dataset contains 1000 rows

Function defination: convert_to_celsius: to convert to unform temperature scale on "Temperature" column

In [8]:
def convert_to_celsius(row):
    if row['Measurement'] == 'F':

        return (row['Temperature'] - 32) * 5.0/9.0
    else:
        return row['Temperature']

Applying convert_to_celsius() on cleaned_df

In [9]:
df_cleaned.loc[:, 'Temperature'] = df_cleaned.apply(lambda row: convert_to_celsius(row), axis=1)
df_cleaned

Cap Shape Cap Color Gill Size Stalk Shape  Habitat  Edibility  \
0      sunken     brown    narrow    tapering   leaves     edible   
1        bell    yellow     broad   enlarging   leaves     edible   
2      convex       red    narrow   enlarging    urban     edible   
3      convex       red     broad    tapering    woods     edible   
4      convex     white    narrow   enlarging    urban     edible   
..        ...       ...       ...         ...      ...        ...   
126   conical     brown    narrow    tapering    urban  poisonous   
127      flat    yellow     broad   enlarging    urban     edible   
128   conical      gray    narrow   enlarging  grasses  poisonous   
129    sunken     white     broad   enlarging    woods  poisonous   
130    convex      gray     broad    tapering   leaves  poisonous   

    Market Demand Spore Print Color Population    Odor  ...  \
0               1            orange   abundant   musty  ...   
1               8             brown   numerous  almond  ...   
2               4            orange   numerous    foul  ...   
3               2            yellow   solitary    foul  ...   
4               1             white    several   musty  ...   
..            ...               ...        ...     ...  ...   
126             5            orange   solitary  almond  ...   
127             4             black   numerous   anise  ...   
128             4            yellow   solitary    foul  ...   
129             9            yellow   abundant    foul  ...   
130             3             black   numerous   anise  ...   

    Stalk Surface Below Ring Stalk Color Above Ring Stalk Color Below Ring  \
0                      silky                  white                  brown   
1                      silky                   gray                   pink   
2                      silky                 orange                 orange   
3                      scaly                  brown                   gray   
4                     smooth                   gray                  white   
..                       ...                    ...                    ...   
126                    scaly                  brown                   pink   
127                    silky                 orange                   pink   
128                   smooth                   gray                   gray   
129                    scaly                   gray                  brown   
130                   smooth                   gray                   gray   

     Veil Type Veil Color         Growth Rate Inspection Date Incubation date  \
0    universal     orange  0.9730964627139668        3/7/2021        3/2/2021   
1    universal     yellow         4.471316648        1/2/2022        1/2/2022   
2      partial      white  3.9783564690608615        3/7/2023        3/3/2023   
3      partial     yellow  1.5922072807712329       4/14/2023        4/5/2023   
4      partial     yellow   3.143234817538799        3/7/2023        3/3/2023   
..         ...        ...                 ...             ...             ...   
126    partial     orange  3.4035697795027953        5/7/2023        5/1/2023   
127    partial      brown  1.9385124867425854       3/11/2023        3/3/2023   
128    partial     yellow  1.4202926693977478        3/2/2023        3/2/2023   
129  universal     orange   3.779518165668043        1/7/2023        1/5/2023   
130    partial      white  1.5888757465434287        3/2/2022        3/1/2022   

    Temperature Measurement  
0          49.0           C  
1          18.0           C  
2           1.0           C  
3           5.0           C  
4           4.0           C  
..          ...         ...  
126        28.0           C  
127        28.0           C  
128        15.0           C  
129         6.0           C  
130         9.0           C  

[131 rows x 23 columns]

Removing "Measurement" column since the unit conversion is done

In [10]:
df_cleaned = df_cleaned.drop('Measurement', axis=1)
df_cleaned

Cap Shape Cap Color Gill Size Stalk Shape  Habitat  Edibility  \
0      sunken     brown    narrow    tapering   leaves     edible   
1        bell    yellow     broad   enlarging   leaves     edible   
2      convex       red    narrow   enlarging    urban     edible   
3      convex       red     broad    tapering    woods     edible   
4      convex     white    narrow   enlarging    urban     edible   
..        ...       ...       ...         ...      ...        ...   
126   conical     brown    narrow    tapering    urban  poisonous   
127      flat    yellow     broad   enlarging    urban     edible   
128   conical      gray    narrow   enlarging  grasses  poisonous   
129    sunken     white     broad   enlarging    woods  poisonous   
130    convex      gray     broad    tapering   leaves  poisonous   

    Market Demand Spore Print Color Population    Odor  ...  \
0               1            orange   abundant   musty  ...   
1               8             brown   numerous  almond  ...   
2               4            orange   numerous    foul  ...   
3               2            yellow   solitary    foul  ...   
4               1             white    several   musty  ...   
..            ...               ...        ...     ...  ...   
126             5            orange   solitary  almond  ...   
127             4             black   numerous   anise  ...   
128             4            yellow   solitary    foul  ...   
129             9            yellow   abundant    foul  ...   
130             3             black   numerous   anise  ...   

    Stalk Surface Above Ring Stalk Surface Below Ring Stalk Color Above Ring  \
0                      scaly                    silky                  white   
1                      silky                    silky                   gray   
2                      silky                    silky                 orange   
3                     smooth                    scaly                  brown   
4                      scaly                   smooth                   gray   
..                       ...                      ...                    ...   
126                    scaly                    scaly                  brown   
127                   smooth                    silky                 orange   
128                  fibrous                   smooth                   gray   
129                   smooth                    scaly                   gray   
130                    scaly                   smooth                   gray   

    Stalk Color Below Ring  Veil Type Veil Color         Growth Rate  \
0                    brown  universal     orange  0.9730964627139668   
1                     pink  universal     yellow         4.471316648   
2                   orange    partial      white  3.9783564690608615   
3                     gray    partial     yellow  1.5922072807712329   
4                    white    partial     yellow   3.143234817538799   
..                     ...        ...        ...                 ...   
126                   pink    partial     orange  3.4035697795027953   
127                   pink    partial      brown  1.9385124867425854   
128                   gray    partial     yellow  1.4202926693977478   
129                  brown  universal     orange   3.779518165668043   
130                   gray    partial      white  1.5888757465434287   

    Inspection Date Incubation date Temperature  
0          3/7/2021        3/2/2021        49.0  
1          1/2/2022        1/2/2022        18.0  
2          3/7/2023        3/3/2023         1.0  
3         4/14/2023        4/5/2023         5.0  
4          3/7/2023        3/3/2023         4.0  
..              ...             ...         ...  
126        5/7/2023        5/1/2023        28.0  
127       3/11/2023        3/3/2023        28.0  
128        3/2/2023        3/2/2023        15.0  
129        1/7/2023        1/5/2023         6.0  
130        3/2/2022        3/1/2022       

Applying StandardScaler on "Growth Rate" & "Temperature" column

In [11]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

df_cleaned_scaled = df_cleaned.copy()
df_cleaned_scaled[['Growth Rate', 'Temperature']] = scaler.fit_transform(df_cleaned[['Growth Rate', 'Temperature']])
df_cleaned_scaled

Cap Shape Cap Color Gill Size Stalk Shape  Habitat  Edibility  \
0      sunken     brown    narrow    tapering   leaves     edible   
1        bell    yellow     broad   enlarging   leaves     edible   
2      convex       red    narrow   enlarging    urban     edible   
3      convex       red     broad    tapering    woods     edible   
4      convex     white    narrow   enlarging    urban     edible   
..        ...       ...       ...         ...      ...        ...   
126   conical     brown    narrow    tapering    urban  poisonous   
127      flat    yellow     broad   enlarging    urban     edible   
128   conical      gray    narrow   enlarging  grasses  poisonous   
129    sunken     white     broad   enlarging    woods  poisonous   
130    convex      gray     broad    tapering   leaves  poisonous   

    Market Demand Spore Print Color Population    Odor  ...  \
0               1            orange   abundant   musty  ...   
1               8             brown   numerous  almond  ...   
2               4            orange   numerous    foul  ...   
3               2            yellow   solitary    foul  ...   
4               1             white    several   musty  ...   
..            ...               ...        ...     ...  ...   
126             5            orange   solitary  almond  ...   
127             4             black   numerous   anise  ...   
128             4            yellow   solitary    foul  ...   
129             9            yellow   abundant    foul  ...   
130             3             black   numerous   anise  ...   

    Stalk Surface Above Ring Stalk Surface Below Ring Stalk Color Above Ring  \
0                      scaly                    silky                  white   
1                      silky                    silky                   gray   
2                      silky                    silky                 orange   
3                     smooth                    scaly                  brown   
4                      scaly                   smooth                   gray   
..                       ...                      ...                    ...   
126                    scaly                    scaly                  brown   
127                   smooth                    silky                 orange   
128                  fibrous                   smooth                   gray   
129                   smooth                    scaly                   gray   
130                    scaly                   smooth                   gray   

    Stalk Color Below Ring  Veil Type Veil Color Growth Rate Inspection Date  \
0                    brown  universal     orange   -1.389191        3/7/2021   
1                     pink  universal     yellow    1.379313        1/2/2022   
2                   orange    partial      white    0.989183        3/7/2023   
3                     gray    partial     yellow   -0.899224       4/14/2023   
4                    white    partial     yellow    0.328265        3/7/2023   
..                     ...        ...        ...         ...             ...   
126                   pink    partial     orange    0.534295        5/7/2023   
127                   pink    partial      brown   -0.625157       3/11/2023   
128                   gray    partial     yellow   -1.035278        3/2/2023   
129                  brown  universal     orange    0.831821        1/7/2023   
130                   gray    partial      white   -0.901861        3/2/2022   

     Incubation date Temperature  
0           3/2/2021    3.520229  
1           1/2/2022    0.436390  
2           3/3/2023   -1.254748  
3           4/5/2023   -0.856833  
4           3/3/2023   -0.956312  
..               ...         ...  
126         5/1/2023    1.431177  
127         3/3/2023    1.431177  
128         3/2/2023    0.137954  
129         1/5/2023   -0.757354  
130         3/1/2022   -0.458918  

[131 rows x 22 columns]

Converting "Market Demand" column to integer data type

In [12]:
df_cleaned_scaled['Market Demand'] = df_cleaned_scaled['Market Demand'].astype(int)

Converting "Inspection Date" &  "Incubation date" to pandas datetime type

In [13]:
df_cleaned_scaled['Inspection Date'] = pd.to_datetime(df_cleaned_scaled['Inspection Date'])

df_cleaned_scaled['Incubation date'] = pd.to_datetime(df_cleaned_scaled['Incubation date'])

Identifying outliers in "Temperature" using Standard(Z-score) (threshold = 3 for simplicity)

In [14]:
z_scores_temperature = np.abs(df_cleaned_scaled['Temperature'])

temperature_outliers = df_cleaned_scaled['Temperature'][z_scores_temperature > 3]
temperature_outliers

0     3.520229
10   -3.012204
Name: Temperature, dtype: float64

Applying Lebel encoding on the following binary cagegorical columns:
1. Gill Size
2. Stalk Shape
3. Edibility
5. Bruising
6. Veil Type

In [15]:
# Identifying columns that are not present in the uploaded dataset based on the user's request for encoding
requested_columns = ['Gill Size', 'Stalk Shape', 'Edibility', 'Bruising', 'Veil Type']
available_columns = [col for col in requested_columns if col in df_cleaned_scaled.columns]

# Reverting to the original dataset for binary encoding on the specified columns
binary_encoded_data = df_cleaned_scaled.copy()

# Applying Label Encoding for binary categorical values on available columns
from sklearn.preprocessing import LabelEncoder
label_encoder = LabelEncoder()

for col in available_columns:
    # Replacing '?' with NaN to handle missing values appropriately
    binary_encoded_data[col] = binary_encoded_data[col].replace('?', pd.NA)
    binary_encoded_data[col] = label_encoder.fit_transform(binary_encoded_data[col].astype(str))

# Display the first few rows of the dataset to inspect the encoded columns
binary_encoded_data[available_columns].head(), available_columns
binary_encoded_data


Cap Shape Cap Color  Gill Size  Stalk Shape  Habitat  Edibility  \
0      sunken     brown          1            1   leaves          0   
1        bell    yellow          0            0   leaves          0   
2      convex       red          1            0    urban          0   
3      convex       red          0            1    woods          0   
4      convex     white          1            0    urban          0   
..        ...       ...        ...          ...      ...        ...   
126   conical     brown          1            1    urban          1   
127      flat    yellow          0            0    urban          0   
128   conical      gray          1            0  grasses          1   
129    sunken     white          0            0    woods          1   
130    convex      gray          0            1   leaves          1   

     Market Demand Spore Print Color Population    Odor  ...  \
0                1            orange   abundant   musty  ...   
1                8             brown   numerous  almond  ...   
2                4            orange   numerous    foul  ...   
3                2            yellow   solitary    foul  ...   
4                1             white    several   musty  ...   
..             ...               ...        ...     ...  ...   
126              5            orange   solitary  almond  ...   
127              4             black   numerous   anise  ...   
128              4            yellow   solitary    foul  ...   
129              9            yellow   abundant    foul  ...   
130              3             black   numerous   anise  ...   

     Stalk Surface Above Ring Stalk Surface Below Ring Stalk Color Above Ring  \
0                       scaly                    silky                  white   
1                       silky                    silky                   gray   
2                       silky                    silky                 orange   
3                      smooth                    scaly                  brown   
4                       scaly                   smooth                   gray   
..                        ...                      ...                    ...   
126                     scaly                    scaly                  brown   
127                    smooth                    silky                 orange   
128                   fibrous                   smooth                   gray   
129                    smooth                    scaly                   gray   
130                     scaly                   smooth                   gray   

    Stalk Color Below Ring Veil Type Veil Color  Growth Rate Inspection Date  \
0                    brown         1     orange    -1.389191      2021-03-07   
1                     pink         1     yellow     1.379313      2022-01-02   
2                   orange         0      white     0.989183      2023-03-07   
3                     gray         0     yellow    -0.899224      2023-04-14   
4                    white         0     yellow     0.328265      2023-03-07   
..                     ...       ...        ...          ...             ...   
126                   pink         0     orange     0.534295      2023-05-07   
127                   pink         0      brown    -0.625157      2023-03-11   
128                   gray         0     yellow    -1.035278      2023-03-02   
129                  brown         1     orange     0.831821      2023-01-07   
130                   gray         0      white    -0.901861      2022-03-02   

     Incubation date Temperature  
0         2021-03-02    3.520229  
1         2022-01-02    0.436390  
2         2023-03-03   -1.254748  
3         2023-04-05   -0.856833  
4         2023-03-03   -0.956312  
..               ...         ...  
126       2023-05-01    1.431177  
127       2023-03-03    1.431177  
128       2023-03-02    0.137954  
129       2023-01-05   -0.757354  
130       2022-03-01   -0.458918  

[131 rows x 22 columns]

Applying One Hot Encoding on the following polytomous variables having more than two values:
1. Cap Shape
2. Cap Color
3. Habitat
4. Spore Print Color
5. Population
6. Odor
7. Ring Type
8. Stalk Surface Above Ring
9. Stalk Surface Below Ring
10. Stalk Color Above Ring
11. Stalk Color Below Ring
12. Veil Color

In [16]:
requested_columns_one_hot = [
    'Cap Shape', 'Cap Color', 'Habitat', 'Spore Print Color', 'Population',
    'Odor', 'Ring Type', 'Stalk Surface Above Ring', 'Stalk Surface Below Ring',
    'Stalk Color Above Ring', 'Stalk Color Below Ring', 'Veil Color'
]

available_columns_one_hot = [col for col in requested_columns_one_hot if col in binary_encoded_data.columns]
encoded_data = pd.get_dummies(binary_encoded_data, columns=available_columns_one_hot)
encoded_data

Gill Size  Stalk Shape  Edibility  Market Demand  Bruising  Veil Type  \
0            1            1          0              1         1          1   
1            0            0          0              8         1          1   
2            1            0          0              4         1          0   
3            0            1          0              2         1          0   
4            1            0          0              1         0          0   
..         ...          ...        ...            ...       ...        ...   
126          1            1          1              5         1          0   
127          0            0          0              4         0          0   
128          1            0          1              4         0          0   
129          0            0          1              9         1          1   
130          0            1          1              3         0          0   

     Growth Rate Inspection Date Incubation date  Temperature  ...  \
0      -1.389191      2021-03-07      2021-03-02     3.520229  ...   
1       1.379313      2022-01-02      2022-01-02     0.436390  ...   
2       0.989183      2023-03-07      2023-03-03    -1.254748  ...   
3      -0.899224      2023-04-14      2023-04-05    -0.856833  ...   
4       0.328265      2023-03-07      2023-03-03    -0.956312  ...   
..           ...             ...             ...          ...  ...   
126     0.534295      2023-05-07      2023-05-01     1.431177  ...   
127    -0.625157      2023-03-11      2023-03-03     1.431177  ...   
128    -1.035278      2023-03-02      2023-03-02     0.137954  ...   
129     0.831821      2023-01-07      2023-01-05    -0.757354  ...   
130    -0.901861      2022-03-02      2022-03-01    -0.458918  ...   

     Stalk Color Above Ring_white  Stalk Color Below Ring_brown  \
0                            True                          True   
1                           False                         False   
2                           False                         False   
3                           False                         False   
4                           False                         False   
..                            ...                           ...   
126                         False                         False   
127                         False                         False   
128                         False                         False   
129                         False                          True   
130                         False                         False   

     Stalk Color Below Ring_gray  Stalk Color Below Ring_orange  \
0                          False                          False   
1                          False                          False   
2                          False                           True   
3                           True                          False   
4                          False                          False   
..                           ...                            ...   
126                        False                          False   
127                        False                          False   
128                         True                          False   
129                        False                          False   
130                         True                          False   

     Stalk Color Below Ring_pink  Stalk Color Below Ring_white  \
0                          False                         False   
1                           True                         False   
2                          False                         False   
3                          False                         False   
4                          False                          True   
..                           ...                           ...   
126                         True                         False   
127                         True                         False   
128                    

# Part-2: Data Visualisation and Communication